### ISSUE TITLE
x86-16 Incorrect output when indexing arrays where index needs type conversion

### ISSUE THREAD

**Issue by blindmatrix** [uxmal/reko#879]
TL;DR; Incorrect code gets generated when accessing arrays where the index needs type conversion for x86-16.

The output for segmented x86-16 code is broken, but I use the output as a template anyways where I can. I usually do this together with a python script that will replace patterns of the output with known symbols from a _.map_ generated with _IDA_. There's a few types of outputs generated but the one I expected for this piece of could was similar to `ds->b011A = (ds->*ds->ptr0094).b0096;` but instead I got `ds->b0118 = (uint16) (uint8) ds->b0094 + 0x95;`.

Sample binaries (and code):
[indexing.zip](https://github.com/uxmal/reko/files/4797837/indexing.zip)

Sample code that demonstrate this bug:

Source code (main.c):
```C
volatile unsigned char index = 0;
volatile unsigned char firstLevel[] = { 1, 2, 3 };
volatile char value;


int main(void) {
	value = firstLevel[index];
	
	return 0;
}
```

Compiled with (Turbo-C 2 under Dosbox):
`tcc -1 -G -ml -f- -eindexing main.c`

Relevant assembly:
```asm
	mov	al,byte ptr DGROUP:_index
	mov	ah,0
	mov	bx,ax
	mov	al,byte ptr DGROUP:_firstLevel[bx]
	mov	byte ptr DGROUP:_value,al
```

After scanning:
```asm
    al = Mem0[ds:0x94:byte]
    ah = 0x00
    bx = ax
    al = Mem0[ds:bx + 0x95:byte]
    Mem0[ds:0x0118:byte] = al
```

After analyzing:
```asm
    Mem11[ds:0x0118:byte] = Mem0[ds:(uint16) (uint8) Mem0[ds:0x94:byte] + 0x95:byte]
```

Finished:
```C
	ds->b0118 = (uint16) (uint8) ds->b0094 + 0x95;
```

Expected:
```C
	ds->b0118 = ds->a0094[ds->b0094]
```
	
If index is unsigned int (`volatile unsigned char index = 0;`) I get _better_ results:
```asm
    bx = Mem0[ds:0x94:word16]
    al = Mem0[ds:bx + 0x96:byte]
    Mem0[ds:0x011A:byte] = al
```

```asm
    Mem8[ds:0x011A:byte] = Mem0[ds:Mem0[ds:0x94:word16] + 0x96:byte]
```

```C
    ds->b011A = (ds->*ds->ptr0094).b0096;
```

**Comment by uxmal** [comment]
Clearly a bug in the Type inference stage. I will try to look at this soon -- but as you know National Herring Day is coming up.

**Comment by blindmatrix** [comment]
Indeed it is. I'll also look around, but I feel this is a bit deeper then what my current understanding for the internals allow....

**Comment by uxmal** [comment]
I have a fix for this issue on my machine. It is causing some regressions in the test suite I need to verify before I can release it.

**Comment by blindmatrix** [comment]
Great! I was looking into something else :P

**Comment by uxmal** [comment]
Thanks for report this. The output from the first example is now:
```C
ds->b0118 = ds->a0094[(uint) (uint8) ds->b0094]
```
(the redundant casts are a separate issue)

**Comment by blindmatrix** [comment]
You are amazing sir ;)
